### PR TITLE
Add support for Python-like negative indexing

### PR DESCRIPTION
**Pull request by brakhane**
As a first step of implementing #4715 I started with negative indexing, since this is
useful in it's own right and a small change so I can get feedback on my coding style before wasting time in implementing something that will never be merged ;)

Negative indexing is a useful feature in Python, especially when combined
with array slicing. Array slicing will hopefully be implemented later, but
negative indexing is useful in it's own right.

A negative index is indexing from the end of an array,
"array[-1] == array[array.size()-1]", using a negative index
larger/smaller than the length of the array is still an error.

While primarily useful for arrays and strings, support is also added to
"array like" structures like Vector3 and Color. This is done just
to be consistent; vector3[2] is much clearer than vector3[-1], but disallowing
it while allowing it for an array with 3 elements seems confusing.

I'm on the fence on that last paragraph, I see no useful use case for it, but think it's better to be consistent and allow negative indexing everywhere. If there is a consensus to only support negative indexing for strings and arrays, I'm fine with changing the PR.

### CODE REVIEW

File: core/variant_op.cpp [hunk 1]
```diff
@@ -2489,110 +2380,14 @@ Variant Variant::get(const Variant& p_index, bool *r_valid) const {
...
+		DEFAULT_OP_ARRAY_CMD(ARRAY, const Array, 0, return (*arr)[index])
+		DEFAULT_OP_DVECTOR_GET(RAW_ARRAY, uint8_t)
+		DEFAULT_OP_DVECTOR_GET(INT_ARRAY, int)
+		DEFAULT_OP_DVECTOR_GET(REAL_ARRAY, real_t)
+		DEFAULT_OP_DVECTOR_GET(STRING_ARRAY, String)
+		DEFAULT_OP_DVECTOR_GET(VECTOR2_ARRAY, Vector2)
+		DEFAULT_OP_DVECTOR_GET(VECTOR3_ARRAY, Vector3)
+		DEFAULT_OP_DVECTOR_GET(COLOR_ARRAY, Color)
 		default: return Variant();
 	}
```
Comment (bojidar-bg) [inline core/variant_op.cpp]: AFAIK, in most of the other code #defines are undefined after they are used. :smile:

Comment (brakhane) [inline core/variant_op.cpp]: Not in variant_op, I wanted to be consistent, but i can add them of course

Comment (bojidar-bg) [inline core/variant_op.cpp]: @brakhane Good then...

**Comment by slapin** [comment]
Well, looks good, but what is our policy regarding documentation?

**Comment by brakhane** [comment]
@slapin Let me guess, it's not "we don't need it"

Joking aside, I couldn't find it, neither in CONTRIBUTING.md nor in the documentation itself.

I suppose I should open a pull request on godot-doc and refer to this pull request?

**Comment by brakhane** [comment]
Added doc to classes.xml, also added example to gdscript tutorial in godot-docs repo

**Comment by reduz** [comment]
my only fear about this is that sometimes people will store unused indices as -1, and having negative index in arrays might result in uncatched bugs. What does everyone else think about this?

**Comment by neikeq** [comment]
@reduz I was thinking about that too. But if python does it like that without problems, I don't see why not.

**Comment by punto-** [comment]
This would be useful, but people sometimes do math to figure out an index,
and a bug might cause the index to be negative, in that case it'd be
preferable to get an error (I'm not sure how python handles this). In my
experience this is used for -1 most of the time, so it might be better to
just have a top() or back() method (maybe with a parameter that defaults to
-1) to get the last element.

**Comment by neikeq** [comment]
or just `array[:-1]`

**Comment by brakhane** [comment]
`array[:-1]` would collide with slicing syntax (which I'm currently working on). I realize that there's no guarantee slicing will make it into GDscript, but if it does, it would be much better to have indexing and slicing have the same syntax as in Python.

Python has had negative indexing for at least 16 years now (probably longer), and experience shows that it solves more problems than it causes. It makes code more readable, and allows you to return an index from a function that says "the last element" without having to know how long the array is.

It is true that negative indexing will cause some code that would error out to now run and do the wrong thing, but the programmer will generally realize that something went wrong and find the bug. Off-by-one errors are already possible, and people cope with it. Again, I can only speak from experience in the Python world that negative indexing is a blessing rather than a curse.

However, changing GDScript behaviour like this should not be done in a point release of 2.0, but added in 2.1 in a "what's new" section to inform people of the new behaviour. But I'm sure that was the plan anyway.

IMHO, using -1 to signify an unintialized index is a bad practise, it's much more clear and safe to just use "null" instead. Indexing with null is still an error. Old code that uses -1 as uninitialized hopefully should have been around long enough that no errorenous indexing is done with them.

**Comment by akien-mga** [comment]
I'm all for using negative indexing, it's a great feature in Python and I never had any issue with it. The use case is also not just `array[-1]`, but things like `array[i-1]` in a loop, etc., and quite importantly array slicing that I'm looking forward to have in GDScript :)

**Comment by reduz** [comment]
okay then, the PR seems fine, so merging it

**Pull request merged**